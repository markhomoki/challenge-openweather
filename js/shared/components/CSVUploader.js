// @flow

import * as React from 'react';
import FormFieldLabel from './FormFieldLabel';

type Props = {
	label?: string,
	onImport: Function,
};

type State = {
	importData: Array<{|date: string, temp: string|}>,
};

export default class CSVUploader extends React.PureComponent<Props, State> {

	static defaultProps = {
		label: 'Select File',
	}

	fileReader: FileReader;

	state = {
		importData: [],
	};

	handleFileRead = () => {
		const content = this.fileReader.result;

		if (!content) {
			return;
		}

		const rows = content.toString().split('\n');
		const res = [];
		for (let i = 0; i < rows.length; i++) {
			const r = rows[i];

			if (r) {
				const cols = r.split(',');

				// Basic check to make sure CSV has the right number of cols and not empty
				if (cols.length === 2 && cols[0] && cols[1]) {
					res.push({
						date: cols[0],
						temp: cols[1],
					});
				}
			}
		}

		this.setState({ importData: res });
	}

	handleFileSelect = (event: SyntheticEvent<*>) => {
		const file = event.target.files[0];

		// Check if file selected
		if (file) {
			this.fileReader = new FileReader();
			this.fileReader.onloadend = this.handleFileRead;
			this.fileReader.readAsText(file);
		}
	}

	import = () => {
		if (this.state.importData.length > 0) {
			this.props.onImport(this.state.importData);
		}
	}

	render() {
		const { label } = this.props;

		return (
			<div className="form-field-file-picker">
				{!!label && <FormFieldLabel>{label}</FormFieldLabel>}
				<input
					type="file"
					accept=".csv"
					onChange={this.handleFileSelect}
				/>
				<button className="button" type="button" onClick={this.import}>Import</button>
			</div>
		);
	}

}
