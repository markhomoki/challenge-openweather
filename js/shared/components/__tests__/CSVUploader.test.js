import React from 'react';
import renderer from 'react-test-renderer';
import CSVUploader from '../CSVUploader';

describe('CSVUploader', () => {
	it('renders correctly', () => {
		const tree = renderer.create(<CSVUploader />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('has different label', () => {
		const wrapper = mount(<CSVUploader label="Choose File" />);
		expect(wrapper.find('.form-field-label').text()).toEqual('Choose File');
	});

});
