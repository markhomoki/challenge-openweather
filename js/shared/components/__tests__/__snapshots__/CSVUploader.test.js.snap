// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSVUploader renders correctly 1`] = `
<div
  className="form-field-file-picker"
>
  <span
    className="form-field-label"
  >
    Select File
  </span>
  <input
    accept=".csv"
    onChange={[Function]}
    type="file"
  />
  <button
    className="button"
    onClick={[Function]}
    type="button"
  >
    Import
  </button>
</div>
`;
